raft/raft_test.go

```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
+	lead := nt.peers[1].(*raft)
+	if lead.prs[3].Match != 1 {
+		t.Fatalf("Node 3 has match %x, want %x", lead.prs[3].Match, 1)
+	}
+
+	// Transfer leader to 3 when node 3 is lack of log.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 3 || lead.transferee != None {
+		t.Fatalf("After transferring, leader is %x transferee %v, want leader 3, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferAfterSnapshot(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+	lead := nt.peers[1].(*raft)
+	nextEnts(lead, nt.storage[1])
+	nt.storage[1].CreateSnapshot(lead.raftLog.applied, &pb.ConfState{Nodes: lead.nodes()}, nil)
+	nt.storage[1].Compact(lead.raftLog.applied)
+
+	nt.recover()
+	if lead.prs[3].Match != 1 {
+		t.Fatalf("Node 3 has match %x, want %x", lead.prs[3].Match, 1)
+	}
+
+	// Transfer leader to 3 when node 3 is lack of snapshot.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	// Only pb.MsgHeartbeatResp can trigger a snapshot.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgHeartbeatResp})
+
+	if lead.lead != 3 || lead.transferee != None {
+		t.Fatalf("After transferring, leader is %x transferee %v, want leader 3, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferToMyselfOrNonExistNode(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to myself, there will be noop.
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgTransferLeader})
+	if lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("After transferring to myself, leader is %x transferee %v, want leader 1, transferee %v", lead.lead, lead.transferee, None)
+	}
+
+	// Transfer leader to non-exist node, there will be noop.
+	nt.send(pb.Message{From: 4, To: 1, Type: pb.MsgTransferLeader})
+	if lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("After transferring to non-exist node, leader is %x transferee %v, want leader 1, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferTimeout(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to isolated node, wait for timeout.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+	for i := 0; i < lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
```

Comment:
t.Fatalf (i do not think we can continue this test if we failed here)

error message: want transferring: transferee = %v, want %v.